SYCL/ESIMD/printf.cpp

```diff
@@ -0,0 +1,121 @@
+#define CONSTANT
+#endif
+
+// This is one of the possible ways to define a format string in a correct
+// address space
+static const CONSTANT char format_hello_world[] = "Hello, World!\n";
+
+// Static isn't really needed if you define it in global scope
+const CONSTANT char format_int[] = "%d\n";
+
+static const CONSTANT char format_vec[] = "%d,%d,%d,%d\n";
+
+const CONSTANT char format_hello_world_2[] = "%lu: Hello, World!\n";
+
+int main() {
+  {
+    queue Queue(esimd_test::ESIMDSelector{},
+                esimd_test::createExceptionHandler());
+
+    Queue.submit([&](handler &CGH) {
+      CGH.single_task<class integral>([=]() {
+        // String
+        oneapi::experimental::printf(format_hello_world);
+        // Due to a bug in Intel CPU Runtime for OpenCL on Windows, information
+        // printed using such format strings (without %-specifiers) might
+        // appear in different order if output is redirected to a file or
+        // another app
+        // FIXME: strictly check output order once the bug is fixed
+        // CHECK: {{(Hello, World!)?}}
+
+        // Integral types
+        oneapi::experimental::printf(format_int, (int32_t)123);
+        oneapi::experimental::printf(format_int, (int32_t)-123);
+        // CHECK: 123
+        // CHECK-NEXT: -123
+
+        // Floating point types
+        {
+          // You can declare format string in non-global scope, but in this case
+          // static keyword is required
+          static const CONSTANT char format[] = "%f\n";
+          oneapi::experimental::printf(format, 33.4f);
+          oneapi::experimental::printf(format, -33.4f);
+        }
+        // CHECK-NEXT: 33.4
+        // CHECK-NEXT: -33.4
+
+        // String types
+        {
+          static CONSTANT const char str_arg[] = "foo";
+          static const CONSTANT char format[] = "%s\n";
+          oneapi::experimental::printf(format, str_arg);
+        }
+        // CHECK-NEXT: foo
+      });
+    });
+    Queue.wait();
+  }
+
+  {
+    queue Queue(esimd_test::ESIMDSelector{},
+                esimd_test::createExceptionHandler());
+    // printf in parallel_for
+    Queue.submit([&](handler &CGH) {
+      CGH.parallel_for<class stream_string>(range<1>(10), [=](id<1> i) {
+        // cast to uint64_t to be sure that we pass 64-bit unsigned value
+        oneapi::experimental::printf(format_hello_world_2, (uint64_t)i.get(0));
+      });
+    });
+    Queue.wait();
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+    // CHECK-NEXT: {{[0-9]+}}: Hello, World!
+  }
+
+  // FIXME: strictly check output order once the bug mentioned above is fixed
```

Comment:
Is it specified in which order the lines should appear?

Reply:
This FIXME is related only to printf without %-specifiers, that is the first case in test.
> // Due to a bug in Intel CPU Runtime for OpenCL on Windows, information
   // printed using such format strings (without %-specifiers) might
   // appear in different order if output is redirected to a file or
   // another app

The idea is to remove last CHECK entirely and modify first one to check output order.

Reply:
I see, That's OK.
Actually, I was asking about CHECKs starting on line 106 whether their order is specified?

Reply:
No, in current configuration is not specified.